Remove func name from errors, etc.

richsalz opened the pull request:
This removes the function names from error messages.  (There is some discussion of this in #9003.)

There are two commits.  The first is all the code changes, the second is the result of doing a force-rebuild of the errors (```make ERROR_REBUILD=-rebuild  errors```) to update the header files and error tables. (*UPDATE* 3 July: there are three commits, one removes unused reason codes.)

For idle minds, this removes nearly 7400 lines.  Comparing "no-shared" versions of openssl:
```
; size apps/openssl.orig apps/openssl
   text	   data	    bss	    dec	    hex	filename
4674788	 290428	  24112	4989328	 4c2190	apps/openssl.orig
4627412	 260412	  24112	4911936	 4af340	apps/openssl
```
Which saves a surprising 75K (1.5%)!

t-j-h commented:
-1 so we can take this to a vote for a decision on this change

davidben commented:
See also #1415.

We did this in BoringSSL years ago. It's worked out very well. In practice, anyone using these were doing so incorrectly. It reduces binary size, reduces maintenance effort, and avoids a large class of backwards compatibility issues where callers incorrectly depend on function codes and break as code moves around internally.

slontis commented:
I think its a good idea - the removed value seems to convey very little.
Would this potentially break things for end users that may be checking these codes?

t-j-h commented:
You have left the function stuff in place for the SYSerr level - so there are still SYS_F_ usages in place.
I was expecting, given the description, that ERR_GET_FUNC (X) would return 0 - but it still returns the function code.

richsalz commented:
I kept those because in this case it shows the actual syscall that failed. If you (or the OMC since you moved for a discussion/vote), prefer that I get rid of them I will.  I suppose, pedantically, the right thing to do is to call ERR_add_error_data with the true syscall name, which I can also do if the OMC wants.

richsalz commented:
Also, if the OMC wants to move forward on this, I would be willing to edit all XXXerr functions to remove the ```_F_``` parameter.  The underlying API, of course, still has to have the function code parameter which it ignores.

levitte commented:
I'm actually unsure how interesting it is to tell what syscall returned with an error.  Isn't the error itself the interesting part?  If, say, `open()` returned with an error, would you naturally code things like this?

``` C
    if ((fd = open("foo.txt", O_CREAT)) < 0)
        perror("Can't create \"foo.txt\"");
```

or like this?

``` C
    if ((fd = open("foo.txt", O_CREAT)) < 0)
        perror("open() doesn't let me create \"foo.txt\"");
```

If we feel the need to display the syscalls that were the origin of an error, maybe that's an indication that would could provide better error messages to start with?  Also, if we absolutely feel the need and go with an error call like `ERR_raise_error` that I mentioned in #9003, it could always be part of additional text for the error.

richsalz commented:
*   I'm actually unsure how interesting it is to tell what syscall returned with an error.

I guess you folks need to decide.  I agree with @t-j-h and find it useful.

levitte commented:
> I agree with @t-j-h

Not sure what you agree with...  the way I read his input was just a state of fact that your works appears incomplete.

richsalz commented:
@levitte, I re-read things and you are right that I misunderstood @t-j-h 's comment.

Unlike other errors however, the ```SYSerr()``` stuff is used to report that something that OpenSSL asked the local system to do failed. It's not self-contained like most of the errors in ```include/openssl/*err.h``` (for example, ```BN_R_PRIVATE_KEY_TOO_LARGE``` and over a thousand others), and it's not like "malloc failure" which we call out explicitly with ```ERR_R_MALLOC_FAILURE``` as one of eight "global errors" in err.h.  Rather the ```SYSerr()``` errors come after a system or library call, and if we don't want our users to dig into the source, we have to at least tell them *what OpenSSL was doing* that failed. This is probably not the final word in better error diagnosis, but it is a (big?) step forward in my view.

And, having written this long and detailed paragraph, of course it belongs over in #9072 but hopefully adding the link will suffice. :)

richsalz commented:
Is anyone on the OMC going to move forward with a discussion/vote on this, or should I close it?

levitte commented:
I just [started the discussion](https://mta.openssl.org/pipermail/openssl-project/2019-June/001426.html)

richsalz commented:
It's been a month since @t-j-h placed the hold/-1 on this.  It seems to have been two weeks since there was discussion on the openssl-project mailing list (https://mta.openssl.org/pipermail/openssl-project/2019-June/001451.html)  Does the @openssl/omc have any plan of action?

If there is no other activity on this topic, I will close this PR and delete the branch next week. I'm fine with that, but I'm not fine with doing the exercise of rebasing all the time if nothing is going to happen.

levitte commented:
That thread is quite long, and I'd like to catch up with it before commenting here (been away for more than a week, I've forgotten things).  This PR is only *part* of the stuff that needs to happen, if any at all.

paulidale commented:
I think removing the function id is a positive change.
@levitte could you elucidate on what else is required?

levitte commented:
>  @levitte could you elucidate on what else is required?

Well, not much that affects *this* PR per se...  we did talk about extending the debug information surrounding an error in that thread, but that should probably be subject for another PR.

levitte commented on crypto/err/err.c:
```diff
@@ -131,9 +101,7 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
      "called a function you should not call"},
     {ERR_R_PASSED_NULL_PARAMETER, "passed a null parameter"},
     {ERR_R_INTERNAL_ERROR, "internal error"},
-    {ERR_R_DISABLED, "called a function that was disabled at compile-time"},
```
We should actually use this more, not take it away...

richsalz replied:
This is a clean-up PR, so leaving something that's never used anywhere in the code seems like a mistake.

levitte replied:
This is a *function code removal PR*, at least according to its description.

levitte replied:
I still want you to undo this...

levitte commented on crypto/err/err.c:
```diff
@@ -131,9 +101,7 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
     {ERR_R_INTERNAL_ERROR, "internal error"},
-    {ERR_R_DISABLED, "called a function that was disabled at compile-time"},
     {ERR_R_INIT_FAIL, "init fail"},
-    {ERR_R_OPERATION_FAIL, "operation fail"},
```
Any reason why you're removing this?

richsalz replied:
I too was kind of surprised that this isn't used anywhere, but it isn't so as part of this cleanup PR I removed it.

levitte replied:
...  and this.

levitte commented on crypto/err/err.c:
```diff
@@ -654,18 +614,9 @@ const char *ERR_lib_error_string(unsigned long e)
-    d.error = ERR_PACK(l, f, 0);
-    p = int_err_get_item(&d);
-    return ((p == NULL) ? NULL : p->string);
+    return ERR_GET_LIB(e) == ERR_LIB_SYS ? "system library" : "internal";
```
I would rather have it return NULL

richsalz replied:
It breaks things if you return NULL.

levitte replied:
Do you mean it breaks `ERR_error_string_n`? Our discussion on printing function indicators might resolve that issue.

richsalz replied:
I know it breaks that function, as I recall while doing some test/dev on this PR something else broke but I don't recall specifics.

levitte replied:
Considering the code before your change *could* return NULL, it seems the callers are assuming too much.

richsalz replied:
changed.

levitte commented on crypto/err/err.c:
```diff
@@ -588,9 +554,9 @@ static unsigned long get_error_values(int inc, int top, const char **file,
 void ERR_error_string_n(unsigned long e, char *buf, size_t len)
 {
-    char lsbuf[64], fsbuf[64], rsbuf[64];
+    char lsbuf[64], rsbuf[64];
```
The generated output should not contain any function indicator at all...  but maybe that's too controversial, so the second option is to simply leave that part of the output empty.

richsalz replied:
I think it's useful to know if it's an openssl function ("internal") or an external fuction ("system) which is what the call to ``ERR_func_error_string``` gives you at line 571.

levitte replied:
That already given with `lsbuf`... I really don't see why the redundancy is needed.

richsalz replied:
There may be applications that parse the colon-separated output and I didn't want to break that.

levitte replied:
That's why I'm suggesting a second option, i.e. leave the function indicator empty.  In other words, you could make this change at line 579:

``` diff
-     BIO_snprintf(buf, len, "error:%08lX:%s:%s:%s", e, ls, fs, rs);
+     BIO_snprintf(buf, len, "error:%08lX:%s::%s", e, ls, rs);
```

levitte replied:
(and similar elsewhere in err.c, where applicable)

richsalz replied:
changed.

richsalz replied:
... and now *every single test* in test_errstr fails because they can't parse the empty double colon, leading to failures like:
```
#   at ../test/recipes/02-test_errstr.t line 109.
/home/rsalz/openssl/github/util/shlib_wrap.sh /home/rsalz/openssl/github/apps/openssl errstr 200001d => 0
not ok 66 - (29) ':Illegal seek' == 'Illegal seek'
```

So I don't want to make that change. I will, however, if @t-j-h's hold/-1 is resolved.

levitte replied:
I'll have a look to see.

levitte replied:
#9304 should help

richsalz replied:
yes, i'll re-do the change once that gets merged.

levitte replied:
#9304 is now merged, ball's back in your court :wink:

richsalz replied:
Done.  Nice.  Next question, is what is necessary to clear @t-j-h's -1/hold?  I don't want to spend any more time on this if it's just going to be wasted effort.

levitte commented on crypto/err/err.c:
```diff
@@ -654,18 +613,9 @@ const char *ERR_lib_error_string(unsigned long e)
-    d.error = ERR_PACK(l, f, 0);
-    p = int_err_get_item(&d);
-    return ((p == NULL) ? NULL : p->string);
+    return ERR_GET_LIB(e) == ERR_LIB_SYS ? "system library" : NULL;
```
No.  Just return NULL.  Having the function indication saying "system library" is redundant and frankly confusing.  *The library indication already says this*

richsalz replied:
Changes being made.  I think keeping ERR_R_DISABLED is a very bad idea, turning compile-time error into possibly run-time.  There's a reason why it was never used :)  But okay it's back.

levitte replied:
It would actually be perfectly suitable in places like this:

https://github.com/openssl/openssl/blob/7a228c391e0a35e1dc1223e3af3371968376857b/crypto/bio/bss_file.c#L420-L423

You know, in place where we *do* signal an error of that type in run-time.

richsalz replied:
Shrug.  It's not used now.  And I believe a more proper fix is to, ya know, not include the function (unless needed for backward compatibilty).  But I am being needlessly pedantic; pax.

richsalz commented:
If the vote passes I'll make additional changes.

levitte commented:
Ok.  I'll just note that other than the few nits, this PR looks good to me.

kroeckx commented:
I think the defines should stay in the public headers for now, should get deprecated, and can all be changed to the value 0.

richsalz commented:
@kroeckx is your comment at https://github.com/openssl/openssl/pull/9058#issuecomment-509007857 enough to clear your -1/hold?  Or do we still need to wait for the vote.  If I change mkerr to output zero-valued ```_F_``` define's, should that block be wrapped in ```#if !OPENSSL_API_3```  or something?

kroeckx commented:
It's @t-j-h that put this PR on hold. (I did this for #9082, you didn't reply to my last comment.)

A vote related to this PR has been called, and is still open. I think it's unclear what the vote really is about, since it actually mentions the PR rather than what you want to achieve. But I think the intention of the vote is for removing the function names from the errors, not on how it's implemented. (I wish people would actually discuss what we voted on before calling a vote to prevent such unclear votes.)

So my comment is about the implementation. If you want, you can consider that a -1 on the implementation that goes away when you make that change.

levitte commented:
> Or do we still need to wait for the vote.

[The vote is ongoing](https://mta.openssl.org/pipermail/openssl-project/2019-July/001465.html), we do need to wait for it either way.

levitte commented:
>  If I change mkerr to output zero-valued `_F_` define's, should that block be wrapped in `#if !OPENSSL_API_3` or something?

I think that's what @kroeckx suggests.  "should get deprecated" looks clear to me.

levitte commented:
> (I wish people would actually discuss what we voted on before calling a vote to prevent such unclear votes.)

Uhmmmm, but there *was* a discussion...  the vote even links to it.  However, mental note made, clearer vote text requested.

levitte commented:
Vote: passed, see https://mta.openssl.org/pipermail/openssl-project/2019-July/001483.html

Note that @kroeckx's desire to keep existing `_F_` macros must be met.

levitte commented:
Oh, and sadly, a rebase is needed.

petrovr commented:
Richard Levitte wrote:
> Vote: passed, seehttps://mta.openssl.org/pipermail/openssl-project/2019-July/001483.html
in absence of replacement functionality .... :(

richsalz commented:
forced-push commits to address all feedback so far.

levitte commented on util/mkerr.pl:
```diff
@@ -500,6 +502,8 @@ sub help
     }
 
     print OUT "\n/*\n * $lib function codes.\n */\n";
+    print OUT "# ifdef OPENSSL_API_3\n";
```
Why is that guard necessary?  ...  Ah, because `#include <openssl/opensslconf.h>` isn't always present.

May I suggest making sure that `#include <openssl/openssl.h>` always be present?

richsalz replied:
okay, done.  revised commits pushed.

levitte commented:
You should probably regenerate the headers in `engines/` as well

richsalz commented:
They are regenerated, in that big commit.

levitte commented:
In that case, `util/mkerr.pl` generates incorrect error macros for the non-internal modules.  I think line 485 is the culprit.

richsalz commented:
> In that case, `util/mkerr.pl` generates incorrect error macros for the non-internal modules. I think line 485 is the culprit.

Okay, updated commit (with some files regenerated) to use ```0``` not the ```f``` parameter.

levitte commented:
I expect Travis to be happier now :wink:

richsalz commented:
Vote passed, all open issues addressed, can this move forward soon?

t8m commented on util/mkerr.pl:
```diff
@@ -122,20 +122,22 @@ sub help
 
 # Data parsed out of the config and state files.
+# We always map function-code values to zero, so items marked below with
+# an asterisk could eventually be removed.  TODO(4.0)
```
So the script should determine whether the header is internal or not and then output the deprecated macro defines or not.

richsalz commented:
We currently do not mark header files (for mkerr) as internal or not.  Do you have a plan to do that?

levitte commented:
> We currently do not mark header files (for mkerr) as internal or not. Do you have a plan to do that?

Those in `include/openssl` are public, anything else is internal.  Does that suffice?

levitte commented:
However, it's not *quite* that simple...  I haven't looked profoundly, but it's possible that we do check some non-public `_F_` macro somewhere.

richsalz commented:
I think removing the ```_F_``` defines in some files goes beyond the scope of this PR.  I removed all internal instances of those macros. I removed them from the header file and then put them back because @kroeckx wanted the compatiblity (that's reasonable).

I really *REALLY* don't want to completely remove them from some headers because we think they're internal-only.  I think Kurt might have an issue with it, too.  Maybe someone can do that as a separate PR or in 4.0

levitte commented:
Yeah, agreed.  They do no harm sitting there...

levitte commented:
Reminder to the second provider: there is a vote result saying we shall do this.

levitte commented:
Merged.

aac96e2797c34a6b2a839eb58c30ab3328a0cee8 Remove function name from errors
cbfa5b03989ee6b8f5c13c4284d5bae02c562f20 Regenerate mkerr files

slontis commented:
So should we do the following from now on?
PROVerr(0, PROV_R_INVALID_QUESTION);

levitte commented:
Yes, that's the idea

t-j-h commented:
Defining all these existing macros to zero and also returning zero as the value of the function value means anyone checking function codes will end up getting false matches once this change is in place.

paulidale commented:
> So should we do the following from now on?
> PROVerr(0, PROV_R_INVALID_QUESTION);

For ones that are new, is it worth dropping the first argument?
For ones that aren't public, is it worth dropping the first argument?
For ones that are public, would an alternate name without the argument be viable?

levitte commented:
> Defining all these existing macros to zero and also returning zero as the value of the function value means anyone checking function codes will end up getting false matches once this change is in place.

Considering the function codes set were a moving target anyway, that risk was tangible either way.

levitte commented:
> For ones that are new, is it worth dropping the first argument?
> For ones that aren't public, is it worth dropping the first argument?
> For ones that are public, would an alternate name without the argument be viable?

So uhmmm, a new name for all, basically?

paulidale commented:
Not for all: `PROVerr`.
I'd not be agin `XXXerror` as a substitute that made them all internal.

richsalz commented:
Adding multiple names is a bad idea; the API surface of OpenSSL is already huge.

Wait until the next major release and rewrite the macros to just drop the argument.

slontis commented:
I dont see any problem with removing it from Proverr which has never been released

levitte commented:
Good point

DDvO commented:
Pleased to see that some simplifications are in progress that will remove redundancy w.r.t. function names and thus the need to keep function names and function code identifiers consistent.

Shouldn't also `crypto/err/openssl.txt` be simplified by removing the "function codes" section?

richsalz commented:
We're keeping the function codes for 3.0 for backward compatibility.  Next release they could be removed; see the TODO comment in util/mkerr.pl

DDvO commented:
I see. So until then, when adding, removing, and renaming functions, one still needs to keep the related function code identifiers in crypto/err/openssl.txt, crypto/*/*err.c, and include/openssl/*err.h in sync.

In particular, when renaming functions that include *err() calls, one still must manually adapt those files (since `util/mkerr.pl` does not support function name changes and instead wrongly assumes new functions are being added and keeps the outdated names).

richsalz commented:
I think it would be easy to change mkerr so that it didn't complain about mismatched function names; would that address your comment?

levitte commented:
New code can simply avoid the function macro completely.

``` C
    FOOerr(0, FOO_R_REASON);
```

(this has been possible since 1.1.0, I might add)

kroeckx commented:
Why isn't the function argument just removed from the call?

kroeckx commented:
I guess to make backporting easier?

richsalz commented:
> Why isn't the function argument just removed from the call?

It's a macro in a public header file, someone may be using it, so we kept it for compatibility for this release, just like we kept the ```_F_``` defines

DDvO commented:
Very glad to learn that for new code one can simply replace the function code identifier argument by 0 in FOOerr(...) calls!
So regarding function codes no need to bother with `mkerr.pl`'s capabilities, either.
 
So we'll do that for our CMP contribution at least for the current chunk 4 and all further ones.
Yet since also chunks 1..3 are not yet released we can prune the function codes there, too, right?
